### PR TITLE
[PLUGIN-1708] Redshift source and connector plugin added.

### PR DESCRIPTION
**Pull request by vikasrathee-cs**
Redshift source and connector plugin added.
https://cdap.atlassian.net/browse/PLUGIN-1708

### CODE REVIEW

File: amazon-redshift-plugin/src/main/java/io/cdap/plugin/amazon/redshift/RedshiftSource.java [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+
+  private final RedshiftSourceConfig redshiftSourceConfig;
+
+  public RedshiftSource(RedshiftSourceConfig redshiftSourceConfig) {
+    super(redshiftSourceConfig);
+    this.redshiftSourceConfig = redshiftSourceConfig;
+  }
+
+  @Override
+  public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
```
Comment (vanathi-g) [inline amazon-redshift-plugin/src/main/java/io/cdap/plugin/amazon/redshift/RedshiftSource.java]: Why is this method override necessary?

File: amazon-redshift-plugin/src/main/java/io/cdap/plugin/amazon/redshift/RedshiftSource.java [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+    return String.format(
+      RedshiftConstants.REDSHIFT_CONNECTION_STRING_FORMAT,
+      redshiftSourceConfig.connection.getHost(),
+      redshiftSourceConfig.connection.getPort(),
+      redshiftSourceConfig.connection.getDatabase());
+  }
+
+  @Override
+  protected LineageRecorder getLineageRecorder(BatchSourceContext context) {
+    String host = redshiftSourceConfig.getConnection().getHost();
```
Comment (vanathi-g) [inline amazon-redshift-plugin/src/main/java/io/cdap/plugin/amazon/redshift/RedshiftSource.java]: nit: can be inlined

File: amazon-redshift-plugin/src/main/java/io/cdap/plugin/amazon/redshift/RedshiftSchemaReader.java [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+import java.util.Set;
+
+/**
+ *    Redshift Schema Reader class
+ */
+public class RedshiftSchemaReader extends CommonSchemaReader {
+
+  private static final Logger LOG = LoggerFactory.getLogger(RedshiftSchemaReader.class);
+
+  public static final Set<Integer> STRING_MAPPED_REDSHIFT_TYPES = ImmutableSet.of(
```
Comment (vanathi-g) [inline amazon-redshift-plugin/src/main/java/io/cdap/plugin/amazon/redshift/RedshiftSchemaReader.java]: Check the documentation here for Java type conversion: https://docs.aws.amazon.com/redshift/latest/mgmt/jdbc20-data-type-mapping.html

Why are we mapping OTHER, ARRAY and SQLXML types? I don't think these are even present in Redshift. Was this tested?

File: amazon-redshift-plugin/src/main/java/io/cdap/plugin/amazon/redshift/RedshiftSchemaReader.java [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+
+    return super.getSchema(metadata, index);
+  }
+
+  @Override
+  public boolean shouldIgnoreColumn(ResultSetMetaData metadata, int index) throws SQLException {
+    if (sessionID == null) {
+      return false;
+    }
+    return metadata.getColumnName(index).equals("c_" + sessionID) ||
```
Comment (vanathi-g) [inline amazon-redshift-plugin/src/main/java/io/cdap/plugin/amazon/redshift/RedshiftSchemaReader.java:94]: What exactly are these columns?

File: amazon-redshift-plugin/src/main/java/io/cdap/plugin/amazon/redshift/RedshiftSource.java (hunk 1 above)
Comment (vikasrathee-cs) [reply amazon-redshift-plugin/src/main/java/io/cdap/plugin/amazon/redshift/RedshiftSource.java]: removed.

File: amazon-redshift-plugin/src/main/java/io/cdap/plugin/amazon/redshift/RedshiftSource.java (hunk 2 above)
Comment (vikasrathee-cs) [reply amazon-redshift-plugin/src/main/java/io/cdap/plugin/amazon/redshift/RedshiftSource.java]: done

File: amazon-redshift-plugin/src/main/java/io/cdap/plugin/amazon/redshift/RedshiftSchemaReader.java (hunk 3 above)
Comment (vikasrathee-cs) [reply amazon-redshift-plugin/src/main/java/io/cdap/plugin/amazon/redshift/RedshiftSchemaReader.java]: Redshift is postgres based DB, some of the array features are moved to Super. We have tested with all data types that redshift allows. Removing this also.

File: amazon-redshift-plugin/src/main/java/io/cdap/plugin/amazon/redshift/RedshiftSchemaReader.java (hunk 4 above)
Comment (vikasrathee-cs) [reply amazon-redshift-plugin/src/main/java/io/cdap/plugin/amazon/redshift/RedshiftSchemaReader.java:94]: This is available in all DB plugins with connector. I have tried to find out its usage, not much documentation is available. So I put it as it is.

File: amazon-redshift-plugin/docs/Redshift-batchsource.md [hunk 5]
```diff
@@ -0,0 +1,104 @@
+# Amazon RedShift Batch Source
+
+
+Description
+-----------
+Reads from an Amazon RedShift database using a configurable SQL query.
```
Comment (yjhawar) [inline amazon-redshift-plugin/docs/Redshift-batchsource.md]: nit : RedShift -> Redshift

File: amazon-redshift-plugin/src/main/java/io/cdap/plugin/amazon/redshift/RedshiftDBRecord.java [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+    int columnType = metadata.getColumnType(columnIndex);
+    if (columnType == Types.NUMERIC) {
+      Schema nonNullableSchema = field.getSchema().isNullable() ?
+        field.getSchema().getNonNullable() : field.getSchema();
+      int precision = metadata.getPrecision(columnIndex);
+      if (precision == 0 && Schema.Type.STRING.equals(nonNullableSchema.getType())) {
+        // When output schema is set to String for precision less numbers
+        recordBuilder.set(field.getName(), resultSet.getString(columnIndex));
+      } else if (Schema.LogicalType.DECIMAL.equals(nonNullableSchema.getLogicalType())) {
+        BigDecimal orgValue = resultSet.getBigDecimal(columnIndex);
```
Comment (yjhawar) [inline amazon-redshift-plugin/src/main/java/io/cdap/plugin/amazon/redshift/RedshiftDBRecord.java]: nit: Any better name for orgValue

File: amazon-redshift-plugin/src/test/java/io/cdap/plugin/amazon/redshift/RedshiftFailedConnectionTest.java [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+ */
+
+package io.cdap.plugin.amazon.redshift;
+
+import io.cdap.plugin.db.connector.DBSpecificFailedConnectionTest;
+import org.junit.Test;
+
+import java.io.IOException;
+
+public class RedshiftFailedConnectionTest extends DBSpecificFailedConnectionTest {
```
Comment (yjhawar) [inline amazon-redshift-plugin/src/test/java/io/cdap/plugin/amazon/redshift/RedshiftFailedConnectionTest.java:24]: Can you mention the reason why the connection fails using redshift jdbc driver but works when using postgrs driver

**Review by yjhawar — APPROVED** [review]
LGTM with minor changes

File: amazon-redshift-plugin/src/test/java/io/cdap/plugin/amazon/redshift/RedshiftFailedConnectionTest.java (hunk 7 above)
Comment (vikasrathee-cs) [reply amazon-redshift-plugin/src/test/java/io/cdap/plugin/amazon/redshift/RedshiftFailedConnectionTest.java:24]: Changed the driver class name in RedshiftConnectorTest to redshift driver. Parent class to both the classes are different and mocking is happening based on that. In this class, it is getting failed due to random values given. In another class it is expecting all the real values and then trying to hit the DB.

**Review by vanathi-g — APPROVED** [review]

File: amazon-redshift-plugin/src/main/java/io/cdap/plugin/amazon/redshift/RedshiftSchemaReader.java (hunk 4 above)
Comment (sau42shri) [reply amazon-redshift-plugin/src/main/java/io/cdap/plugin/amazon/redshift/RedshiftSchemaReader.java:94]: @vikasrathee-cs The PR is approved. Can this be merged now?

Comment (vikasrathee-cs) [reply amazon-redshift-plugin/src/main/java/io/cdap/plugin/amazon/redshift/RedshiftSchemaReader.java:94]: yes Saurabh, merging it.

**Pull request merged**